modules/tensor_mechanics/src/kernels/StressDivergenceRZTensors.C

```diff
@@ -0,0 +1,127 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "StressDivergenceRZTensors.h"
+
+template<>
+InputParameters validParams<StressDivergenceRZTensors>()
+{
+  InputParameters params = validParams<StressDivergenceTensors>();
+  params.addClassDescription ("Calculate stress divergence for an axisymmetric problem in cylinderical coordinates.");
+  params.addRequiredParam<unsigned int>("component", "An integer corresponding to the direction the variable this kernel acts in. (0 for x, 1 for y, 2 for z; note in this kernel disp_x refers to the radial displacement and disp_y refers to the axial displacement.)");
+  params.set<bool>("use_displaced_mesh") = true;
+  return params;
+}
+
+StressDivergenceRZTensors::StressDivergenceRZTensors(const std::string & name, InputParameters parameters) :
+    StressDivergenceTensors(name, parameters),
+    _q_point(_assembly.qPoints())
+{
+}
+
+Real
+StressDivergenceRZTensors::computeQpResidual()
+{
+  Real div(0);
+  if (_component == 0)
+  {
+    div = _grad_test[_i][_qp](0) * _stress[_qp](0,0) +
+        + ( _test[_i][_qp] / _q_point[_qp](0)) * _stress[_qp](2,2) +
+        + _grad_test[_i][_qp](1) * _stress[_qp](0,1); // stress_{rz}
+//      + 0 * _stress[_qp](1,1)
+  }
+  else if (_component == 1)
+  {
+    div = _grad_test[_i][_qp](1) * _stress[_qp](1,1) +
+        + _grad_test[_i][_qp](0) * _stress[_qp](1,0);  // stress_{zr}
+//      + 0 * _stress[_qp](0,0) + 0 * _stress[_qp](2,2)
+  }
+  else
+    mooseError("Invalid component for this AxisymmetricRZ problem.");
+
+  return div;
+}
+
+Real
+StressDivergenceRZTensors::computeQpJacobian()
+{
+  return calculateJacobian(_component, _component);
+}
+
+Real
+StressDivergenceRZTensors::computeQpOffDiagJacobian(unsigned int jvar)
+{
+  if (_xdisp_coupled && jvar == _xdisp_var)
+    return calculateJacobian( _component, 0 );
+
+  else if (_ydisp_coupled && jvar == _ydisp_var)
+    return calculateJacobian( _component, 1 );
+
+  else if (_temp_coupled && jvar == _temp_var)
+    return 0.0;
+
+  return 0;
+}
+
+Real
+StressDivergenceRZTensors::calculateJacobian( unsigned int ivar, unsigned int jvar)
```

Comment:
I think it would improve clarity of the code if you just moved the calculations directly into the ```computeQpJacobian``` and ```computeQpOffDiagJacobian``` methods. ```computeQpOffDiagJacobian``` never calculates and on-diagonal contribution so you are not avoiding any code duplication by using this method.